pkg/sources/sourcer/sourcer.go

```diff
@@ -0,0 +1,51 @@
+/*
+Copyright 2022 The Numaproj Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sourcer
+
+import (
+	"context"
+	"io"
+
+	"github.com/numaproj/numaflow/pkg/forwarder"
+	"github.com/numaproj/numaflow/pkg/isb"
+)
+
+type SourceReader interface {
+	io.Closer
+	// GetName returns the name of the source.
+	GetName() string
+	// Read reads a chunk of messages and returns at the first occurrence of an error. Error does not indicate that the
+	// array of result is empty, the callee should process all the elements in the array even if the error is set. Read
+	// will not mark the message in the buffer as "READ" if the read for that index is erring.
+	// There is a chance that we have read the message and the container got forcefully terminated before processing. To provide
+	// at-least-once semantics for reading, during the restart we will have to reprocess all unacknowledged messages.
+	Read(context.Context, int64) ([]*isb.ReadMessage, error)
+	// Ack acknowledges an array of offset.
+	Ack(context.Context, []isb.Offset) []error
+	// NoAck cancels acknowledgement of an array of offset.
+	NoAck(context.Context, []isb.Offset)
+	// Partitions returns the partitions of the source.
+	Partitions() []int32
```

Comment:
Can we add more descriptions to this method? I have some questions about it. 1. could you help me confirm my understanding that calling this method multiple times can return different results (source partition list can dynamically change)? and 2. with this change, all udsources need to implement it too, what's the benefit of introducing this method?

Reply:
1. Yes it can result in different results for sources like Kafka which involves partition rebalancing when a pod comes up or gets deleted.
2. Yes, it should not be a complex operation since we already expect the users to provide partition information in the offset. The reason is when we are publishing idle watermarks we cannot publish them to all the partitions instead we should only publish idle watermarks to the partitions which are assigned/managed by this pod.

Reply:
+1. Thanks for the explanation!